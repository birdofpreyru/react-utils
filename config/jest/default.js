/* eslint-disable import/no-commonjs */
/* global __dirname, module, process, require */

const path = require('node:path');

const globalLibDir = path.resolve(process.execPath, '../../lib/node_modules');

module.exports = {
  collectCoverage: true,
  collectCoverageFrom: [
    'src/**/*.{js,jsx,ts,tsx}',
  ],
  coverageDirectory: '__coverage__',
  moduleDirectories: [
    'node_modules',
    globalLibDir,
  ],
  moduleNameMapper: {
    '\\.(scss|css)$': 'identity-obj-proxy',
  },
  modulePathIgnorePatterns: ['<rootDir>/build/'],
  resolver: `${__dirname}/resolver.js`,
  rootDir: '../..',
  setupFilesAfterEnv: [
    '<rootDir>/config/jest/setup.js',
  ],
  testEnvironmentOptions: {
    url: 'http://localhost',
  },
  testMatch: ['**/__tests__/**/*.(j|t)s?(x)'],
  testPathIgnorePatterns: [
    '/__assets__/',
    '/node_modules/',
  ],
  transform: {
    '\\.((j|t)sx?|svg)$': 'babel-jest',
  },
  transformIgnorePatterns: [
    '/node_modules/(?!@dr.pogodin/react-utils|uuid)',
  ],

};
