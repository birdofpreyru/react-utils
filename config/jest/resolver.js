/* global module */

module.exports = (path, options) => {
  const ops = { ...options };

  // Appending "development" option we ensure that Jest tests use development
  // version of the library's build for browsers, which for example contains
  // "data-testid" attributes (they are optimized-out of production builds).
  if (ops.conditions && !ops.conditions.includes('development')) {
    ops.conditions = [...ops.conditions, 'development'];
  }

  return options.defaultResolver(path, ops);
};
