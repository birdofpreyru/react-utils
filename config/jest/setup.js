/* global globalThis, require */

// eslint-disable-next-line import/no-unassigned-import
import 'raf/polyfill';

// TODO: This is a temporary polyfill necessary for react-router,
// as JSDom does not provide TextEncoder, see:
// https://github.com/remix-run/react-router/issues/12363
if (!globalThis.TextEncoder || !globalThis.TextDecoder) {
  const { TextDecoder, TextEncoder } = require('node:util');
  globalThis.TextEncoder = TextEncoder;
  globalThis.TextDecoder = TextDecoder;
}
