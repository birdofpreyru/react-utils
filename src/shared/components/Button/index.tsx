// The <Button> component implements a standard button / button-like link.

import type {
  FunctionComponent,
  KeyboardEventHandler,
  MouseEventHandler,
  PointerEventHandler,
  ReactNode,
} from 'react';

import themed, { type Theme } from '@dr.pogodin/react-themes';

import Link from 'components/Link';

import defaultTheme from './style.scss';

type PropsT = {
  active?: boolean;
  children?: ReactNode;
  disabled?: boolean;
  enforceA?: boolean;
  onClick?: MouseEventHandler & KeyboardEventHandler;
  onKeyDown?: KeyboardEventHandler;
  onKeyUp?: KeyboardEventHandler;
  onMouseDown?: MouseEventHandler;
  onMouseUp?: MouseEventHandler;
  onPointerDown?: PointerEventHandler;
  onPointerUp?: PointerEventHandler;
  openNewTab?: boolean;
  replace?: boolean;
  testId?: string;
  theme: Theme<'active' | 'button' | 'disabled'>;
  // TODO: It needs a more precise typing of the object option.
  to?: object | string;
};

export const BaseButton: FunctionComponent<PropsT> = ({
  active,
  children,
  disabled,
  enforceA,
  onClick,
  onKeyDown: onKeyDownProp,
  onKeyUp,
  onMouseDown,
  onMouseUp,
  onPointerDown,
  onPointerUp,
  openNewTab,
  replace,
  testId,
  theme,
  to,
}) => {
  let className = theme.button;
  if (active && theme.active) className += ` ${theme.active}`;
  if (disabled) {
    if (theme.disabled) className += ` ${theme.disabled}`;
    return (
      <div
        className={className}
        data-testid={process.env.NODE_ENV === 'production' ? undefined : testId}
      >
        {children}
      </div>
    );
  }

  let onKeyDown = onKeyDownProp;
  if (!onKeyDown && onClick) {
    onKeyDown = (e) => {
      if (e.key === 'Enter') onClick(e);
    };
  }

  if (to) {
    return (
      <Link
        className={className}
        data-testid={process.env.NODE_ENV === 'production' ? undefined : testId}
        enforceA={enforceA}
        onClick={onClick}

        // TODO: For now, the `onKeyDown` handler is not passed to the <Link>,
        // as <Link> component does not call it anyway, presumably due to
        // the inner implementation details. We should look into supporting it:
        // https://github.com/birdofpreyru/react-utils/issues/444
        // onKeyDown={onKeyDown}

        onKeyUp={onKeyUp}
        onMouseDown={onMouseDown}
        onMouseUp={onMouseUp}
        onPointerDown={onPointerDown}
        onPointerUp={onPointerUp}
        openNewTab={openNewTab}
        replace={replace}
        to={to}
      >
        {children}
      </Link>
    );
  }

  return (
    <div
      className={className}
      data-testid={process.env.NODE_ENV === 'production' ? undefined : testId}
      onClick={onClick}
      onKeyDown={onKeyDown}
      onKeyUp={onKeyUp}
      onMouseDown={onMouseDown}
      onMouseUp={onMouseUp}
      onPointerDown={onPointerDown}
      onPointerUp={onPointerUp}
      role="button"
      tabIndex={0}
    >
      {children}
    </div>
  );
};

/**
 * Button component theme: a map of CSS
 * class names to append to button elements:
 * @prop {string} [active] to the root element of active button.
 * @prop {string} [button] to the root element of any button.
 * @prop {string} [disabled] to the root element of disabled button.
 */
export default themed(BaseButton, 'Button', defaultTheme);
