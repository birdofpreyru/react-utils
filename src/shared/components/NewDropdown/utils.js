/**
 * Auxiliary functions for <Dropdown> component.
 */

import { findIndex, findLastIndex } from 'lodash';

/**
 * Returns option name and value as two elements of array.
 * @param {object|string} option
 * @return {string[]}
 */
export function optionNameValue(option) {
  return typeof option === 'string'
    ? [option, option] : [option.name, option.value];
}

/**
 * Returns option value.
 * @param {object|string} option Option object or string.
 * @return {string}
 */
export function optionValue(option) {
  return typeof option === 'string' ? option : option.value;
}

/**
 * Finds the option by value.
 * @param {string} value Option value.
 * @param {array} options Options array.
 * @param {function} [filter] Optional. Filter function.
 * @return {number} Index of the matching option, or -1.
 */
function findOptionIndex(value, options, filter) {
  return options.findIndex(
    (item) => optionValue(item) === value && (!filter || filter(item)),
  );
}

/**
 * Finds the next option index.
 * @param {string} value Current value.
 * @param {array} options Option array.
 * @param {function} [filter] Optional. Filter function.
 * @return {number} Next option index, or -1.
 */
export function findNextOptionIndex(value, options, filter) {
  let index = findOptionIndex(value, options, filter);
  if (index >= 0 && ++index < options.length) {
    return filter ? findIndex(options, filter, index) : index;
  }
  return -1;
}

/**
 * Finds the previous option index.
 * @param {string} value Current value.
 * @param {array} options Option array.
 * @param {function} [filter] Optional. Filter function.
 * @return {number} Previous option index, or -1.
 */
export function findPrevOptionIndex(value, options, filter) {
  let index = findOptionIndex(value, options, filter) - 1;
  if (index >= 0 && filter) index = findLastIndex(options, filter, index);
  return index >= 0 ? index : -1;
}
