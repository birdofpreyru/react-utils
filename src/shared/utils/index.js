import themed, {
  COMPOSE,
  PRIORITY,
  ThemeProvider,
} from '@dr.pogodin/react-themes';

import config from './config';
import * as isomorphy from './isomorphy';
import time from './time';
import * as url from './url';
import * as webpack from './webpack';

themed.COMPOSE = COMPOSE;
themed.PRIORITY = PRIORITY;

let JU = null; // eslint-disable-line import/no-mutable-exports
if (isomorphy.IS_SERVER_SIDE
  && (process.env.NODE_CONFIG_ENV || process.env.NODE_ENV) !== 'production'
) {
  /* eslint-disable global-require */
  const path = webpack.requireWeak('path');
  JU = webpack.requireWeak(path.resolve(__dirname, './jest'));
  /* eslint-enable global-require */
}

/**
 * Creates a new async barrier: a Promise instance with its resolve method
 * attached as .resolve field.
 * @return {Promise<>}
 */
function newBarrier() {
  let resolve;
  const promise = new Promise((res) => { resolve = res; });
  promise.resolve = resolve;
  return promise;
}

/**
 * Attempts to execute provided async "action" up to "maxRetries" time with
 * the given "interval". If any of attempts is successful it returns its result,
 * otherwise if all attempts fail it throws the error of last attempt.
 * @param {function} action
 * @param {number} [maxRetries=5] Optional. Maximum number of retries. Defaults
 *  to 5 attempts.
 * @param {number} [interval=1000] Optional. Interval between retries [ms].
 *  Defaults to 1 second.
 * @return {Promise} Resolves to the result of successful operation, or
 *  rejects with the error from the latst failed attempt.
 */
async function withRetries(action, maxRetries = 5, interval = 1000) {
  /* eslint-disable no-await-in-loop */
  for (let n = 1; ; ++n) {
    try {
      return await action();
    } catch (error) {
      if (n < maxRetries) await time.timer(interval);
      else throw error;
    }
  }
  /* eslint-enable no-await-in-loop */
}

export {
  config,
  isomorphy,
  JU,
  newBarrier,
  themed,
  ThemeProvider,
  time,
  url,
  webpack,
  withRetries,
};
