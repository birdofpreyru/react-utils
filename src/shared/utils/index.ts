import themedImpl, {
  COMPOSE,
  PRIORITY,
  type Theme,
  ThemeProvider,
} from '@dr.pogodin/react-themes';

import config from './config';
import * as isomorphy from './isomorphy';
import time from './time';
import * as webpack from './webpack';

export {
  type Listener,
  Barrier,
  Emitter,
  Semaphore,
  withRetries,
} from '@dr.pogodin/js-utils';

export { getSsrContext } from './globalState';
export { default as splitComponent } from './splitComponent';

type ThemedT = typeof themedImpl & {
  COMPOSE: typeof COMPOSE;
  PRIORITY: typeof PRIORITY;
};

const themed: ThemedT = themedImpl as ThemedT;

themed.COMPOSE = COMPOSE;
themed.PRIORITY = PRIORITY;

export {
  type Theme,
  config,
  isomorphy,
  themed,
  ThemeProvider,
  time,
  webpack,
};
