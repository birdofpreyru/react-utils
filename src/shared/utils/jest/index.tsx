/* global jest, document */
/* eslint-disable import/no-extraneous-dependencies */

import type {
  AxiosRequestConfig,
  AxiosResponse,
  AxiosStatic,
  InternalAxiosRequestConfig,
} from 'axios';

import mockdate from 'mockdate';
import { type ReactNode, act } from 'react';
import { type Root, createRoot } from 'react-dom/client';

import { type RenderResult, render } from '@testing-library/react';

/**
 * An alias for [act(..)](https://reactjs.org/docs/test-utils.html#act)
 * from `react`.
 */
export { act };

export { default as getGlobal } from './global';

global.IS_REACT_ACT_ENVIRONMENT = true;

const originalProcessVersions = process.versions;

/**
 * Tricks **react-utils** into thinking the test is running within client-side
 * (browser) environment.
 */
export function mockClientSide() {
  Object.defineProperty(process, 'versions', { value: undefined });
}

/**
 * Reverts the effect of {@link module:JU.mockClientSide mockClientSide(..)}.
 */
export function unmockClientSide() {
  Object.defineProperty(process, 'versions', {
    value: originalProcessVersions,
    writable: false,
  });
}

/**
 * Generates a mock UUID, or better said it determenistically transforms given
 * `seed` number into a UUID-formatted string.
 * @param {number} seed
 * @return {string}
 */
export function getMockUuid(seed = 0) {
  const x = seed.toString(16).padStart(32, '0');
  return `${x.slice(0, 8)}-${x.slice(8, 12)}-${x.slice(12, 16)}-${x.slice(16, 20)}-${x.slice(20)}`;
}

export type AxiosRequestHandlerT =
  (config: AxiosRequestConfig) => Partial<AxiosResponse> | null | undefined;

export function mockAxios(handlers: AxiosRequestHandlerT[]) {
  const axios: AxiosStatic = jest.requireActual('axios');

  axios.defaults.adapter = async (
    config: AxiosRequestConfig,
  ): Promise<AxiosResponse> => {
    for (const handler of handlers) {
      const res = handler(config);
      if (res) {
        return {
          config: config as InternalAxiosRequestConfig,
          data: null,
          headers: {},
          status: 200,
          statusText: 'OK',
          ...res,
        };
      }
    }

    // Fallback to the regular network request.
    let res: AxiosResponse;
    try {
      res = await axios({ ...config, adapter: ['xhr', 'http', 'fetch'] });
      console.warn(
        'Network request has not been mocked for a test.\n\nConfig:\n',
        config,
        '\n\nResult:\n',
        JSON.stringify(res, null, 2),
      );
    } catch (e) {
      console.warn(
        'Network request has not been mocked for a test, and failed.\n\nConfig:\n',
        config,
        '\n\nError\n',
        JSON.stringify(e, null, 2),
      );
      throw e;
    }

    return res;
  };

  return axios;
}

/**
 * Advances mock timers, and mock date by the specified time.
 * @param {number} time Time step [ms].
 * @returns {Promise} Wait for this to "jump after" any async code which should
 * be executed because of the mock time movement.
 */
export async function mockTimer(time: number) {
  mockdate.set(time + Date.now());
  await jest.advanceTimersByTimeAsync(time);
}

export type MountedSceneT = HTMLElement & {
  destroy: () => void;
  snapshot: () => void;
};

/**
 * Mounts `scene` to the DOM, and returns the root scene element.
 * @param scene
 * @return Created container DOM element with destroy() function
 *  attached.
 */
export function mount(scene: ReactNode): MountedSceneT {
  let root: Root;
  const element = document.createElement('div');
  document.body.appendChild(element);

  const res: MountedSceneT = (element as unknown) as MountedSceneT;
  res.destroy = () => {
    // NOTE: As it seems @testing-library may reset this flag to false
    // when it is simulating user events.
    global.IS_REACT_ACT_ENVIRONMENT = true;

    act(() => root.unmount());
    res.remove();
  };

  res.snapshot = () => {
    expect(res).toMatchSnapshot();
  };

  // NOTE: As it seems @testing-library may reset this flag to false
  // when it is simulating user events.
  global.IS_REACT_ACT_ENVIRONMENT = true;

  act(() => {
    root = createRoot(res);
    root.render(scene);
  });
  return res;
}

type SnapshotOptionsT = {
  await?: Promise<void>;
};

export async function snapshot(
  element: React.ReactElement,
  options?: SnapshotOptionsT,
) {
  let res: RenderResult | undefined;

  const promise = act(() => {
    res = render(element);
    return options?.await;
  });

  if (res === undefined) throw Error('Render failed');
  if (options?.await) await promise;

  const nodes = res.asFragment().childNodes;
  expect(nodes.length > 1 ? [...nodes] : nodes[0]).toMatchSnapshot();
  return res;
}
