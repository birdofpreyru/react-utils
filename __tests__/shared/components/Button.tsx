/** @jest-environment jsdom */

import { noop, omit } from 'lodash';
import { StaticRouter } from 'react-router';

import Button from 'components/Button';
import { snapshot } from 'utils/jest';

const testTheme = {
  active: 'ACTIVE_CLASS',
  ad: 'AD_TAG',
  button: 'BUTTON_CLASS',
  context: 'CONTEXT_TAG',
  disabled: 'DISABLED_CLASS',
  hoc: 'HOC_TAG',
};

describe('Matches snapshots', () => {
  test('when active', () => {
    snapshot((
      <Button
        active
        onClick={noop}
        theme={testTheme}
      >
        BUTTON
      </Button>
    ));
  });

  test('when disabled', () => {
    snapshot((
      <Button
        disabled
        onClick={noop}
        theme={testTheme}
      >
        BUTTON
      </Button>
    ));
    snapshot((
      <Button
        disabled
        onClick={noop}
        theme={omit(testTheme, 'disabled')}
      >
        BUTTON
      </Button>
    ));
  });

  test('when rendered as link', () => {
    snapshot((
      <StaticRouter location="">
        <Button
          active
          theme={testTheme}
          to="/SOME/TEST/URL"
        >
          BUTTON
        </Button>
      </StaticRouter>
    ));
  });
});
