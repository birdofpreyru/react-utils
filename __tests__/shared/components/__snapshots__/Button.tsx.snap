// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Matches snapshots when active 1`] = `
<div
  class="-dr-pogodin-react-utils___src-shared-components-Button-style___button___E1FNQT AD_TAG HOC_TAG BUTTON_CLASS -dr-pogodin-react-utils___src-shared-components-Button-style___active___MAe9O6 AD_TAG HOC_TAG ACTIVE_CLASS"
  role="button"
  tabindex="0"
>
  BUTTON
</div>
`;

exports[`Matches snapshots when disabled 1`] = `
<div
  class="-dr-pogodin-react-utils___src-shared-components-Button-style___button___E1FNQT AD_TAG HOC_TAG BUTTON_CLASS -dr-pogodin-react-utils___src-shared-components-Button-style___disabled___Br9IWV AD_TAG HOC_TAG DISABLED_CLASS"
>
  BUTTON
</div>
`;

exports[`Matches snapshots when disabled 2`] = `
<div
  class="-dr-pogodin-react-utils___src-shared-components-Button-style___button___E1FNQT AD_TAG HOC_TAG BUTTON_CLASS -dr-pogodin-react-utils___src-shared-components-Button-style___disabled___Br9IWV"
>
  BUTTON
</div>
`;

exports[`Matches snapshots when rendered as link 1`] = `
<a
  class="-dr-pogodin-react-utils___src-shared-components-Button-style___button___E1FNQT AD_TAG HOC_TAG BUTTON_CLASS -dr-pogodin-react-utils___src-shared-components-Button-style___active___MAe9O6 AD_TAG HOC_TAG ACTIVE_CLASS"
  href="/SOME/TEST/URL"
>
  BUTTON
</a>
`;
